File: dijkstra.js
var Dijkstra = function(start_point, end_point, r ){

    var start = start_point, 
          end = end_point; 

    var output = []; 

    //create a matrix
    var createMatrix = function(rows, columns) {

              var mat = new Array( rows );

              for (var i = 0; i < rows; i++) 

                      mat[i] = new Array( columns );
         return mat;
    };

    //just declare and create an array and fill them with a for loop
    var createArray = function( n ) {
 
        var arr = new Array( n )

        for(var i = 0; i< arr.length; i++) 

            arr[i] = 0;

        return arr; 
    }  

    var matrix = r,

        Road = createMatrix(6, 6),

        R = createArray( 6 ),

        F = createArray( 6 ),

        S = createArray( 6 ),

        nodes = 0,

        get = function( node ) {

              if(F[node]) get(F[node])

              output.push(node)
        },

        getCost = function() {
                return (R[end]) 
        },

        getShortestPath = function() {
                             
             get( end );

             return output;
        }, 

        read = (function() {

               var n = r.length, x, y, cost;

               for(var i = 1; i <= 5; i++) {

                       for(var j = 1; j <= 5; j++) {

                           if(i == j) Road[i][j] = 0; 
                               else   Road[i][j] = 88; 
                       }
               }

               for(var i = 0; i < n; i++) {

                   x = matrix[i][0]; 
                   y = matrix[i][1];

                   cost = matrix[i][2];
                   Road[x][y] = cost
               } 
        })(),

        solve = (function() {

                var min, posMin;

                S[start] = 1;
                 
                for(var i = 1; i <= 5; i++) {

                    R[ i ] = Road[ start ][ i ];

                    if( start != i ) 

                        if( R[ i ] < 88) 

                            F[ i ] = start
                }

                //execute n-1 times
                for(var i=1;i<=5-1;i++) {
     
                    min = 88

                    for(var k = 1; k <= 5; k++) {

                        if(S[ k ] == 0) 

                           if(R[ k ] < min) {

                              min = R[ k ]

                              posMin = k
                           }  
                    }

                    //marked the node as selected
                    S[ posMin ] = 1

                    for(var j = 1; j <= 5; j++) {

                        if(S[j] == 0) 

                           if(R[ j ] > R[ posMin ] + Road[ posMin ][ j ]) {

                                     R[ j ] = R[ posMin ] + Road[ posMin ][ j ]
                                     F[ j ] = posMin
                           }  
                    }
                }

        })();
        
     
        return {
               getCost: getCost,
               getShortestPath: getShortestPath
        
        }
}

File: dijkstra.rb
class Dijkstra

      def initialize(filename, startpoint, endpoint)

         #start node
         @start = startpoint
          
         #end node
         @end = endpoint

         #in this vector we will hold the shortest path 
         @path = []
 
         #PInfinit
         @PInfinit = 88

         #read and init the father array, selected array and road array
         readAndInit(filename)

         #Dijkstra's algorithm in action and good luck
         dijkstra()
         
      end  

      # This method determines the minimum cost of the shortest path
      def getCost()
          @R[@end] 
      end

      # get the shortest path
      def getShortestPath()
          ROAD(@end)
          @path
      end

      def ROAD(node) 
          if @F[node] != 0
             ROAD(@F[node])
          end 

          @path.push(node)
      end

      def dijkstra()
  
          start = @start
          min = @PInfinit
          posMin = @PInfinit

          (1..@nodes-1).each do |i|
              @R[i] = @road[start][i]
              if i != start 
                 if @R[i] < @PInfinit
                    @F[i] = start 
                 end
              end
          end
  
          #for debug 
          #print @R    
  
          @S[start] = 1

          #for debug
          #print @S

          (1..@nodes-2).each do |d|
               
              min = @PInfinit

              (1..@nodes-1).each do |i|
                  if @S[i] == 0
                     if @R[i] < min
                         min = @R[i]
                         posMin = i
                     end
                  end  
              end     

              @S[posMin] = 1  

              (1..@nodes-1).each do|j|
                  if @S[j] == 0
                           if @R[j] > @R[posMin] + @road[posMin][j] 
                              @R[j] = @R[posMin] + @road[posMin][j]
                              @F[j] = posMin
                           end 
                  end
              end  
               
          end 

      end 

      def readAndInit(file)

          arr = []

          File.open(file, "r").each_line { |line| arr << line.split(' ').map {|c| c.to_i} }

          @nodes = arr[0][0] + 1

          n = arr.size()-1

          @road = Array.new(@nodes) { Array.new(@nodes) }
          @R = Array.new(@nodes)
          @S = Array.new(@nodes)
          @F = Array.new(@nodes)

          (0..@nodes-1).each do |i|
              @R[i] = 0 
          end

          (0..@nodes-1).each do |i|
              @S[i] = 0 
          end

          (0..@nodes-1).each do |i|
              @F[i] = 0 
          end

          (0..@nodes-1).each do |i|
              (0..@nodes-1).each do |j|
                  if i == j
                     @road[i][j] = 0 
                  else
                     @road[i][j] = @PInfinit
                  end 
              end  
           end 
  
           (1..n).each do |i|
                  x = arr[i][0]
                  y = arr[i][1]
                  c = arr[i][2]
                  @road[x][y] = c
           end 
          
      end
end

start_point = 1
end_point = 3

ob = Dijkstra.new('dijkstra.in', start_point, end_point)

print "Cost = ", ob.getCost(), "\n"
print "Shortest Path = ", ob.getShortestPath()